pkg/tracing/tracing.go

```diff
@@ -0,0 +1,95 @@
+package tracing
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc/credentials"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// NewTraceConfig generates the initial tracing configuration with 'endpoint' as the endpoint to connect to the Opentelemetry Collector
+func NewTraceConfig(endpoint string, certs string, kubeClient kubernetes.Interface, log logr.Logger) error {
+	ctx := context.Background()
+
+	var client otlptrace.Client
+
+	if certs != "" {
+		// here the certificates are stored as configmaps
+		secret, err := kubeClient.CoreV1().Secrets("kyverno").Get(ctx, certs, v1.GetOptions{})
+		if err != nil {
+			log.Error(err, "Error fetching certificate from secret")
+		}
+
+		cp := x509.NewCertPool()
+		if !cp.AppendCertsFromPEM([]byte(secret.Data["ca.pem"])) {
+			return fmt.Errorf("credentials: failed to append certificates")
+		}
+
+		transportCreds := credentials.NewClientTLSFromCert(cp, "")
+
+		client = otlptracegrpc.NewClient(
+			otlptracegrpc.WithEndpoint(endpoint),
+			otlptracegrpc.WithTLSCredentials(transportCreds),
+		)
+	} else {
+		client = otlptracegrpc.NewClient(
+			otlptracegrpc.WithEndpoint(endpoint),
+			otlptracegrpc.WithInsecure(),
+		)
+	}
```

Comment:
Duplicate code, can we extract it to a separate function and re-use it?

Reply:
I was thinking a function ```FetchCert``` in a utils package?

Reply:
Needed for both tracing and metrics